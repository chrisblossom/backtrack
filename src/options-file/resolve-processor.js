/* @flow */

import ResolveWithPrefix from 'resolve-with-prefix';
import path from 'path';
import { toArray } from '../utils/object-utils';

import type { Resolve } from '../types';

const resolve = new ResolveWithPrefix();

type Args = {|
    value: Resolve | $ReadOnlyArray<Resolve>,
    current?: Resolve,
    dirname: string,
|};

function resolveProcessor(args: Args): Resolve {
    const { value, current = {}, dirname } = args;

    const normalizeValue = toArray(value);

    const result = normalizeValue.reduce((acc, packageId) => {
        let id;
        let pkgId;

        if (typeof packageId === 'string') {
            id = packageId;
            pkgId = packageId;
        } else if (packageId.packageId) {
            id = packageId.id;
            pkgId = packageId.packageId;
        } else if (packageId.packagePath) {
            id = packageId.id;
            pkgId = path.resolve(dirname, packageId.packagePath);
        } else {
            throw new Error('invalid packageId');
        }

        const resolved = resolve(pkgId, { dirname });

        return {
            ...acc,
            [id]: resolved,
        };
    }, current);

    return result;
}

export { resolveProcessor };
