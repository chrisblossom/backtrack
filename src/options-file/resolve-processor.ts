import path from 'path';
import { resolveWithPrefixSync } from 'resolve-with-prefix';
import { toArray } from '../utils/object-utils';
import { Resolve } from '../types';

interface Args {
	value: Resolve | readonly Resolve[];
	current?: Resolve;
	dirname: string;
}

function resolveProcessor(args: Args): Resolve {
	const { value, current = {}, dirname } = args;

	const normalizeValue: Resolve[] = toArray(
		// @ts-ignore
		value,
	);

	const result = normalizeValue.reduce((acc, packageId) => {
		let id;
		let pkgId;

		if (typeof packageId === 'string') {
			id = packageId;
			pkgId = packageId;
		} else if (packageId.packageId) {
			id = packageId.id;
			pkgId = packageId.packageId;
		} else if (packageId.packagePath) {
			id = packageId.id;
			pkgId = path.resolve(dirname, packageId.packagePath);
		} else {
			throw new Error('invalid packageId');
		}

		const resolved = resolveWithPrefixSync(pkgId, { dirname });

		return {
			...acc,
			[id]: resolved,
		};
	}, current);

	return result;
}

export { resolveProcessor };
