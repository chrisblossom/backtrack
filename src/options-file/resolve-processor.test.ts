import path from 'path';

const resolveProcessor = (args: any) =>
	require('./resolve-processor').resolveProcessor(args);

describe('resolveProcessor', () => {
	const cwd = process.cwd();

	afterEach(() => {
		process.chdir(cwd);
	});

	test('resolves multiple presets', () => {
		const value = [
			'backtrack-preset-01',
			'backtrack-preset-02',
		];
		const dirname = path.resolve(__dirname, '__sandbox__/app1/');
		process.chdir(dirname);

		const result = resolveProcessor({
			value,
			dirname,
		});

		expect(result).toMatchSnapshot();
	});

	test('allow for packageId', () => {
		const value = [{ id: 'renamed', packageId: 'backtrack-preset-01' }];
		const dirname = path.resolve(__dirname, '__sandbox__/app1/');
		process.chdir(dirname);

		const result = resolveProcessor({
			value,
			dirname,
		});

		expect(result).toMatchSnapshot();
	});

	test('allow for packagePath', () => {
		const value = [
			{
				id: 'preset-01',
				packagePath: path.normalize('./preset-01.js'),
			},
		];

		const dirname = path.resolve(__dirname, '__sandbox__/');
		process.chdir(dirname);

		const result = resolveProcessor({
			value,
			dirname,
		});

		expect(result).toMatchSnapshot();
	});

	test('allows single', () => {
		const value = 'backtrack-preset-01';
		const dirname = path.resolve(__dirname, '__sandbox__/app1/');
		process.chdir(dirname);

		const result = resolveProcessor({
			value,
			dirname,
		});

		expect(result).toMatchSnapshot();
	});

	test('adds to current', () => {
		const value = ['backtrack-preset-01'];
		const dirname = path.resolve(__dirname, '__sandbox__/app1/');
		process.chdir(dirname);

		const current = {
			eslint: path.normalize('/path/to/eslint'),
		};

		const result = resolveProcessor({
			value,
			current,
			dirname,
		});

		expect(result).toMatchSnapshot();
	});

	test('overwrites key', () => {
		const value = ['backtrack-preset-01'];
		const dirname = path.resolve(__dirname, '__sandbox__/app1/');
		process.chdir(dirname);

		const current = {
			'backtrack-preset-01': path.normalize(
				'/path/to/backtrack-preset-01',
			),
		};

		const result = resolveProcessor({
			value,
			current,
			dirname,
		});

		expect(result).toMatchSnapshot();
	});
});
