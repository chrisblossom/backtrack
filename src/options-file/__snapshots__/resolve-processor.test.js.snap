// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveProcessor adds to current 1`] = `
Object {
  "backtrack-preset-01": "<PROJECT_ROOT>/node_modules/backtrack-preset-01/index.js",
  "eslint": "/path/to/eslint",
}
`;

exports[`resolveProcessor allow for packageId 1`] = `
Object {
  "renamed": "<PROJECT_ROOT>/node_modules/backtrack-preset-01/index.js",
}
`;

exports[`resolveProcessor allow for packagePath 1`] = `
Object {
  "preset-01": "<PROJECT_ROOT>/preset-01.js",
}
`;

exports[`resolveProcessor allows single 1`] = `
Object {
  "backtrack-preset-01": "<PROJECT_ROOT>/node_modules/backtrack-preset-01/index.js",
}
`;

exports[`resolveProcessor overwrites key 1`] = `
Object {
  "backtrack-preset-01": "<PROJECT_ROOT>/node_modules/backtrack-preset-01/index.js",
}
`;

exports[`resolveProcessor resolves multiple presets 1`] = `
Object {
  "backtrack-preset-01": "<PROJECT_ROOT>/node_modules/backtrack-preset-01/index.js",
  "backtrack-preset-02": "<PROJECT_ROOT>/node_modules/backtrack-preset-02/index.js",
}
`;
